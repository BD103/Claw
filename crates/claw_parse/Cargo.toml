[package]
name = "claw_parse"
version = "0.1.0-dev"
authors = ["BD103"]
edition = "2021"
description = "Programming language that compiles into Scratch blocks (SB3)."
readme = "README.md"
# homepage = "https://bd103.github.io/Claw"
repository = "https://github.com/BD103/Claw"
license = "MIT OR Apache-2.0"
publish = false

[dependencies]
ariadne = "0.1.5"
chumsky = "0.8.0"

# Required for snapshot testing
serde = { version = "1.0.145", optional = true, features = ["derive"] }

[dev-dependencies]
insta = { version = "1.20.0", features = ["glob", "ron"] }

[[test]]
name = "snap"
path = "tests/snap.rs"
required-features = ["serde"]
