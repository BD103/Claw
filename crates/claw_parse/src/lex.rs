use claw_middle::Span;
use logos::Logos;

pub fn lex<'source>(source: &'source str) -> Tokenizer<'source> {
    Tokenizer::new(source)
}

pub struct Tokenizer<'source> {
    inner: logos::Lexer<'source, TokenKind>,
}

impl<'source> Tokenizer<'source> {
    fn new(source: &'source str) -> Self {
        Tokenizer {
            inner: TokenKind::lexer(source),
        }
    }
}

impl<'source> Iterator for Tokenizer<'source> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        Some(Token {
            kind: self.inner.next()?,
            span: self.inner.span(),
        })
    }
}

pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

#[derive(Logos, Debug, PartialEq)]
#[non_exhaustive]
pub enum TokenKind {
    #[error]
    Error,
}
