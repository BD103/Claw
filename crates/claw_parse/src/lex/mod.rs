mod token;

pub use self::token::{Token, TokenKind, token};

pub fn lex<'a>(source: &'a str) -> Tokenizer<'a> {
    Tokenizer::new(source)
}

pub struct Tokenizer<'a> {
    cursor: usize,
    source: &'a str,
}

impl<'a> Tokenizer<'a> {
    fn new(source: &'a str) -> Self {
        Tokenizer {
            cursor: 0,
            source,
        }
    }
}

impl Iterator for Tokenizer<'_> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        if self.cursor > self.source.len() - 1 {
            return Some(Token {
                kind: token!(EOF),
                span: 0..0,
            });
        }

        todo!()
    }
}
