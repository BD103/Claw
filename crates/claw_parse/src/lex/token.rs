use std::ops::Range;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Token {
    pub kind: TokenKind,
    pub span: Range<usize>,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub enum TokenKind {
    // Values
    /// Either an identifier or keyword.
    ///
    /// `fn`, `looks`, etc.
    Ident(String),
    /// A string.
    ///
    /// `"Some text"`
    Text(String),
    /// A number, optional decimal and sign.
    ///
    /// `64`, `-2`, `3.1415`
    Number(String),
    /// True or false.
    ///
    /// `true` or `false`
    Boolean(bool),

    // Delimiters
    /// `(`
    ParenOpen,
    /// `)`
    ParenClose,
    /// `{`
    BracketOpen,
    /// `}`
    BracketClose,

    // Separators
    /// `::`
    DoubleColon,
    /// `:`
    Colon,
    /// `;`
    Semi,
    /// `,`
    Comma,

    // Prefixes
    /// `@`
    AtSign,
    /// `$`
    EnumSign,

    // Comparison
    /// `==`
    Eq,
    /// `!=`
    NotEq,
    /// `&&`
    And,
    /// `||`
    Or,
    /// `>`
    Gt,
    /// `>=`
    Ge,
    /// `<`
    Lt,
    /// `<=`
    Le,
}
