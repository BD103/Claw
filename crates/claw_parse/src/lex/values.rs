use super::TokenKind;
use crate::error::LexError;
use chumsky::prelude::*;

pub fn create_ident() -> impl Parser<char, TokenKind, Error = LexError> {
    text::ident().map(TokenKind::Ident)
}

pub fn create_text() -> impl Parser<char, TokenKind, Error = LexError> {
    just('"')
        .ignore_then(filter(|c| *c != '"').repeated())
        .then_ignore(just('"'))
        .collect::<String>()
        .map(TokenKind::Text)
}

pub fn create_number() -> impl Parser<char, TokenKind, Error = LexError> {
    let num = text::int(10)
        .chain::<char, _, _>(just('.').chain(text::digits(10)).or_not().flatten())
        .collect::<String>();

    // Supports negatives
    choice((
        just('-').then(num.clone()).map(|(_, mut x)| {
            x.insert(0, '-');
            TokenKind::Number(x)
        }),
        num.map(TokenKind::Number),
    ))
}

pub fn create_bool() -> impl Parser<char, TokenKind, Error = LexError> {
    text::keyword("true")
        .ignored()
        .to(TokenKind::Boolean(true))
        .or(text::keyword("false")
            .ignored()
            .to(TokenKind::Boolean(false)))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ident() {
        let input = [
            "hi",
            "hiThere",
            "hi_there",
            "_hiThErE",
            "h1Th3r3",
        ];
        // Invalid idents
        let input_err = [
            "5ithere",
            "-hello",
            "ğŸ¦€",
        ];

        let parser = create_ident();

        input.into_iter()
            .for_each(|x| {
                let output = parser.parse(x);
                assert_eq!(output, Ok(TokenKind::Ident(x.into())));
            });
        
        input_err.into_iter()
            .for_each(|x| {
                let output = parser.parse(x);
                assert!(output.is_err());
            });
        
    }
}
