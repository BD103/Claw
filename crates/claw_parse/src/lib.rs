#![doc = include_str!("../README.md")]

pub mod ast;
pub mod error;
pub mod lex;
pub mod parse;

use ariadne::Report;
use chumsky::Parser;

pub fn parse(script: &str) -> Result<(), Box<Report>> {
    let lexer = lex::create_lexer();
    let tokens = lexer.parse(script);

    let tokens = match tokens {
        Ok(tokens) => tokens,
        Err(errors) => {
            return Err(Box::new(error::generate_token_report(errors)));
        }
    };

    let parser = parse::create_parser();
    let ast = parser.parse(tokens);

    match ast {
        Ok(ast) => {
            dbg!(ast);
        },
        Err(errors) => {
            dbg!(errors);
        },
    };

    Ok(())
}
