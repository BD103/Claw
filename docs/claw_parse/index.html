<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="`claw_parse`"><meta name="keywords" content="rust, rustlang, rust-lang, claw_parse"><title>claw_parse - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../dark.css" disabled><link rel="stylesheet" type="text/css" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script defer src="../crates.js"></script><script defer src="../main.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../claw_parse/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location"></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../claw_parse/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location"><a href="#">Crate claw_parse</a></h2><div class="sidebar-elems"><div class="block"><ul><li class="version">Version 0.1.0-dev</li><li><a id="all-types" href="all.html">All Items</a></li></ul></div><section><div class="block"><ul><li><a href="#reexports">Re-exports</a></li><li><a href="#modules">Modules</a></li><li><a href="#functions">Functions</a></li><li><a href="#types">Type Definitions</a></li></ul></div></section></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../claw_parse/index.html"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><button type="button">?</button></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div></div></form></nav></div><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn"><span class="in-band">Crate <a class="mod" href="#">claw_parse</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../src/claw_parse/lib.rs.html#1-25">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><h2 id="claw_parse"><a href="#claw_parse"><code>claw_parse</code></a></h2>
<p>This crate is part of the Claw programming language. You can see the project README <a href="https://github.com/BD103/Claw">here</a>.</p>
</div></details><h2 id="reexports" class="small-section-header"><a href="#reexports">Re-exports</a></h2><div class="item-table"><div class="item-row"><div class="item-left import-item" id="reexport.AST"><code>pub use crate::ast::<a class="type" href="ast/type.AST.html" title="type claw_parse::ast::AST">AST</a>;</code></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left import-item" id="reexport.create_parser"><code>pub use crate::parser::<a class="fn" href="parser/fn.create_parser.html" title="fn claw_parse::parser::create_parser">create_parser</a>;</code></div><div class="item-right docblock-short"></div></div></div><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="mod" href="ast/index.html" title="claw_parse::ast mod">ast</a></div><div class="item-right docblock-short"><p>Contains the abstract syntax tree definitions.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="parser/index.html" title="claw_parse::parser mod">parser</a></div><div class="item-right docblock-short"><p>Contains the logic of the actual parser, written using [<code>chumsky</code>].</p>
</div></div></div><h2 id="functions" class="small-section-header"><a href="#functions">Functions</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.build_report.html" title="claw_parse::build_report fn">build_report</a></div><div class="item-right docblock-short"><p>Transforms a result to use [<code>ariadne</code>]’s [<code>Report</code>].</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.get_source.html" title="claw_parse::get_source fn">get_source</a></div><div class="item-right docblock-short"><p>Returns the [<code>Source</code>] of a string reference.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.parse.html" title="claw_parse::parse fn">parse</a></div><div class="item-right docblock-short"><p>Parses a <a href="https://doc.rust-lang.org/1.64.0/std/primitive.str.html" title="str"><code>str</code></a> into <a href="ast/type.AST.html" title="AST"><code>AST</code></a>.</p>
</div></div></div><h2 id="types" class="small-section-header"><a href="#types">Type Definitions</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="type" href="type.ParseError.html" title="claw_parse::ParseError type">ParseError</a></div><div class="item-right docblock-short"><p>Error type returned by the parsers.</p>
</div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="claw_parse" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.64.0 (a55dd71d5 2022-09-19)" ></div></body></html>