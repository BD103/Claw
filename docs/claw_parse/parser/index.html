<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Contains the logic of the actual parser, written using [`chumsky`]."><meta name="keywords" content="rust, rustlang, rust-lang, parser"><title>claw_parse::parser - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../../dark.css" disabled><link rel="stylesheet" type="text/css" href="../../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../../storage.js"></script><script defer src="../../main.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../../favicon.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../../claw_parse/index.html"><div class="logo-container"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></div></a><h2 class="location"></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../../claw_parse/index.html"><div class="logo-container"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></div></a><h2 class="location"><a href="#">Module parser</a></h2><div class="sidebar-elems"><section><div class="block"><ul><li><a href="#functions">Functions</a></li></ul></div></section></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../../claw_parse/index.html"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><button type="button">?</button></div><div id="settings-menu" tabindex="-1"><a href="../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../wheel.svg"></a></div></div></form></nav></div><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn"><span class="in-band">Module <a href="../index.html">claw_parse</a>::<wbr><a class="mod" href="#">parser</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../../src/claw_parse/parser/mod.rs.html#1-75">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Contains the logic of the actual parser, written using [<code>chumsky</code>].</p>
</div></details><h2 id="functions" class="small-section-header"><a href="#functions">Functions</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.create_call.html" title="claw_parse::parser::create_call fn">create_call</a></div><div class="item-right docblock-short"><p>Creates a call parser.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.create_comment.html" title="claw_parse::parser::create_comment fn">create_comment</a></div><div class="item-right docblock-short"><p>Creates a comment parser.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.create_declare.html" title="claw_parse::parser::create_declare fn">create_declare</a></div><div class="item-right docblock-short"><p>Creates an enum declaration parser.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.create_expression.html" title="claw_parse::parser::create_expression fn">create_expression</a></div><div class="item-right docblock-short"><p>Creates an expression parser.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.create_func.html" title="claw_parse::parser::create_func fn">create_func</a></div><div class="item-right docblock-short"><p>Creates a function declaration parser.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.create_parser.html" title="claw_parse::parser::create_parser fn">create_parser</a></div><div class="item-right docblock-short"><p>Creates the main [<code>Parser</code>] to be used when parsing a Claw script.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.create_sprite.html" title="claw_parse::parser::create_sprite fn">create_sprite</a></div><div class="item-right docblock-short"><p>Creates a sprite declaration parser.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.create_statement.html" title="claw_parse::parser::create_statement fn">create_statement</a></div><div class="item-right docblock-short"><p>Creates a statement parser.</p>
</div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../../" data-current-crate="claw_parse" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.64.0 (a55dd71d5 2022-09-19)" ></div></body></html>