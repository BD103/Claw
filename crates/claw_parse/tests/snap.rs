// If this isn't running when you execute test, make sure to enable the "snap" feature. This will be
// skipped if disabled. https://doc.rust-lang.org/stable/cargo/reference/cargo-targets.html#the-required-features-field

use insta::{assert_ron_snapshot, glob};
use std::fs::read_to_string;

#[test]
fn lexer() {
    use chumsky::Parser;
    use claw_parse::lex::*;

    let lexer = create_lexer();

    glob!("samples/*.claw", |path| {
        let res = lexer.parse(read_to_string(path).unwrap());

        assert!(res.is_ok());
        assert_ron_snapshot!(res.unwrap());
    });
}
