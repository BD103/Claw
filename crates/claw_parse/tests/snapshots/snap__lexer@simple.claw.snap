---
source: crates/claw_parse/tests/snap.rs
expression: res.unwrap()
input_file: crates/claw_parse/tests/samples/simple.claw
---
[
  Token(
    kind: AtSign,
    span: Range(
      start: 0,
      end: 1,
    ),
  ),
  Token(
    kind: Ident("on"),
    span: Range(
      start: 1,
      end: 3,
    ),
  ),
  Token(
    kind: ParenOpen,
    span: Range(
      start: 3,
      end: 4,
    ),
  ),
  Token(
    kind: Ident("start"),
    span: Range(
      start: 4,
      end: 9,
    ),
  ),
  Token(
    kind: ParenClose,
    span: Range(
      start: 9,
      end: 10,
    ),
  ),
  Token(
    kind: Ident("fn"),
    span: Range(
      start: 11,
      end: 13,
    ),
  ),
  Token(
    kind: Ident("do_thing"),
    span: Range(
      start: 14,
      end: 22,
    ),
  ),
  Token(
    kind: ParenOpen,
    span: Range(
      start: 22,
      end: 23,
    ),
  ),
  Token(
    kind: ParenClose,
    span: Range(
      start: 23,
      end: 24,
    ),
  ),
  Token(
    kind: BracketOpen,
    span: Range(
      start: 25,
      end: 26,
    ),
  ),
  Token(
    kind: Ident("looks"),
    span: Range(
      start: 109,
      end: 114,
    ),
  ),
  Token(
    kind: DoubleColon,
    span: Range(
      start: 114,
      end: 116,
    ),
  ),
  Token(
    kind: Ident("show"),
    span: Range(
      start: 116,
      end: 120,
    ),
  ),
  Token(
    kind: ParenOpen,
    span: Range(
      start: 120,
      end: 121,
    ),
  ),
  Token(
    kind: ParenClose,
    span: Range(
      start: 121,
      end: 122,
    ),
  ),
  Token(
    kind: Semi,
    span: Range(
      start: 122,
      end: 123,
    ),
  ),
  Token(
    kind: BracketClose,
    span: Range(
      start: 124,
      end: 125,
    ),
  ),
]
